pkg/compactor/blocks_cleaner.go

```diff
@@ -397,6 +413,46 @@ func (c *BlocksCleaner) deleteUserMarkedForDeletion(ctx context.Context, userID
 	return nil
 }
 
+// withRetries invokes the given function as many times as it takes according to
+// the backoff config. Each invocation of f will be given perCallTimeout to
+// complete. This is specifically designed to retry timeouts due to flaky
+// connectivity with the objstore backend.
+func withRetries(ctx context.Context, perCallTimeout time.Duration, bc backoff.Config, logger log.Logger, f func(context.Context) error) error {
```

Comment:
I like this functional approach towards doing retries. Maybe something like this could go into dskit? Alternatively, we could use [Failsafe-go](https://failsafe-go.dev/retry/#usage) for this kind of functional approach to retries.

Reply:
I was just looking at failsafe-go yesterday, randomly. Whoever wrote that has been around the block a few times. :)

> Maybe something like this could go into dskit?

Yeah, I looked to see if it already existed in dskit. There's probably room for something there. For this application, the coupling of per-call timeout contexts and a shouldRetry function that looks for `DeadlineExceeded` seemed a little single-purpose.

Failsafe's backoff looks nice. But there is a plus to sticking with dskit/backoff as it is so pervasive in this codebase.

I expect to come back into this file to add similar retries inside of UpdateIndex ("sometime") so maybe we can keep this dialogue open/rule of three and all that?